backends-common/postgres/src/main/java/org/apache/james/backends/postgres/utils/PostgresExecutor.java

```diff
@@ -139,19 +156,21 @@ public Mono<Boolean> executeExists(Function<DSLContext, SelectConditionStep<?>>
     }
 
     public Mono<Integer> executeReturnAffectedRowsCount(Function<DSLContext, Mono<Integer>> queryFunction) {
-        return dslContext()
-            .flatMap(queryFunction)
-            .retryWhen(Retry.backoff(MAX_RETRY_ATTEMPTS, MIN_BACKOFF)
-                .filter(preparedStatementConflictException()));
+        return Mono.usingWhen(jamesPostgresConnectionFactory.getConnection(domain),
+            connection -> dslContext(connection)
+                .flatMap(queryFunction)
+                .retryWhen(Retry.backoff(MAX_RETRY_ATTEMPTS, MIN_BACKOFF)
+                    .filter(preparedStatementConflictException())),
+            jamesPostgresConnectionFactory::closeConnection);
     }
 
     public Mono<Connection> connection() {
-        return connection;
+        return jamesPostgresConnectionFactory.getConnection(domain);
     }
 
     @VisibleForTesting
     public Mono<Void> dispose() {
-        return connection.flatMap(con -> Mono.from(con.close()));
+        return jamesPostgresConnectionFactory.close();
```

Comment:
Where will invoke this method?

If no invoke, we will early get an error: "out of pool"

Reply:
I release connection in PostgresExecutor

Reply:
Got your idea
It looks nice 
don't forget close connection in method `public Mono<Connection> connection() {`